File: solutions/csharp/roman-numerals/1/RomanNumerals.cs
using System.Collections.Generic;
using System.Text;

public static class RomanNumeralExtension
{
    private static readonly Dictionary<int, string> Numerals = new ()
    {
        { 1000, "M" }, { 900, "CM" },
        { 500,  "D" }, { 400, "CD" },
        { 100,  "C" }, { 90,  "XC" },
        { 50,   "L" }, { 40,  "XL" },
        { 10,   "X" }, { 9,   "IX" },
        { 5,    "V" }, { 4,   "IV" },
        { 1,    "I" }
    };

    public static string ToRoman(this int value)
    {
        var roman = new StringBuilder();

        foreach (var key in Numerals.Keys)
        {
            var count = value / key;

            for (var i = 0; i < count; i++)
                roman.Append(Numerals[key]);

            value %= key;
        }

        return roman.ToString();
    }
}

File: solutions/lua/roman-numerals/1/roman-numerals.lua
return {
  to_roman = function(n)
    local numerals = {
      [1000] = "M", [900] = "CM",
      [ 500] = "D", [400] = "CD",
      [ 100] = "C", [ 90] = "XC",
      [  50] = "L", [ 40] = "XL",
      [  10] = "X", [  9] = "IX",
      [   5] = "V", [  4] = "IV",
      [   1] = "I"
    }

    -- Sort by key (descending) for calculations
    local keys = {}

    for k, _ in pairs(numerals) do
      table.insert(keys, k)
    end

    table.sort(keys, function(a, b) return a > b end)

    -- Build roman numeral
    local roman = ""

    for i = 1, #keys do
      local key   = keys[i]
      local count = n // key

      for j = 1, count do
        roman = roman .. numerals[key]
      end

      n = n % key
    end

    return roman
  end
}
